### PR TITLE
PT-234 Instant Alerts Onboarding Popup

### PR DESCRIPTION
**Pull request by asugar13**
### Description:
This PR implements the instant alerts onboarding pop up modal. It will be opened when clicked on the component implemented in this related PR (https://github.com/Financial-Times/n-myft-ui/pull/590)

For now, this is just a modal that matches designs. The modal has been added to the demo. We will be implementing the functionality to open/close the modal on a subsequent PR.

[JIRA](https://financialtimes.atlassian.net/jira/software/c/projects/PT/boards/1655?modal=detail&selectedIssue=PT-234)

Designs: https://www.figma.com/file/yyojM73dWDAF5BPgrMv7du/PT-219-To-increase-the-number-of-instant-alert-users?type=design&node-id=703-6064&mode=design&t=UvW4KTXefFngMMCs-0

Screenshots:

https://github.com/Financial-Times/n-myft-ui/assets/10453619/01df058f-e209-4d1f-8beb-c4761a889d55

### CODE REVIEW

File: components/instant-alerts-onboarding/instant-alerts-onboarding.html [hunk 1]
```diff
@@ -0,0 +1,23 @@
+{{#if @root.flags.myFtApiWrite}}
+	<div class="n-myft-ui__instant-alerts-onboarding">
+		<div class="n-myft-ui__instant-alerts-onboarding__content">
+			<form>
+				<span class="o-forms-input o-forms-input--checkbox n-myft-ui__instant-alerts-onboarding__checkbox">
+					<label for="receive-instant-alerts">
+						<input type="checkbox" name="receive-instant-alerts" value="receive-instant-alerts">
```
Comment (asugar13) [inline components/instant-alerts-onboarding/instant-alerts-onboarding.html:7]: **Note**: a11y tests are not passing because of this checkbox. I need to figure out a way to submit the form by clicking on the checkbox.

File: components/instant-alerts-onboarding/main.scss [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+.n-myft-ui__instant-alerts-onboarding__list-preferences {
+	margin-bottom: 0;
+	font-size: 14px;
+}
+
+.n-myft-ui__instant-alerts-onboarding__manage-preferences {
+	margin-top: 0;
+	margin-bottom: 20px;
+	// font-size: 14px;
```
Comment (asugar13) [inline components/instant-alerts-onboarding/main.scss:32]: **Note**: I left the `font-size` commented out to highlight the fact that there are a few design inconsistencies in the font choices.

File: components/instant-alerts-onboarding/instant-alerts-onboarding.html (hunk 1 above)
Comment (GlynnPhillips) [reply components/instant-alerts-onboarding/instant-alerts-onboarding.html:7]: **Suggestion:** A possible suggestion here (I am still thinking if its the best way) would be to have hidden submit button within the form. Then when the user clicks on the checkbox we can submit the form via JS instead of using the submit button.

File: components/instant-alerts-onboarding/main.scss (hunk 2 above)
Comment (GlynnPhillips) [reply components/instant-alerts-onboarding/main.scss:32]: Can you provide more details about what you mean by this?

File: components/instant-alerts-onboarding/instant-alerts-onboarding.html [hunk 3]
```diff
@@ -0,0 +1,23 @@
+{{#if @root.flags.myFtApiWrite}}
+	<div class="n-myft-ui__instant-alerts-onboarding">
```
Comment (GlynnPhillips) [inline components/instant-alerts-onboarding/instant-alerts-onboarding.html:2]: **Suggestion:** I wonder if our naming could be more generic / future proofed? Maybe something like `n-myft-ui__follow-menu` or `n-myft-ui__follow-manage` so that if we want to add other options into here we can?

File: components/instant-alerts-onboarding/main.scss [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	border-radius: 10px;
+	border: 2px solid oColorsByName('black-5');
+	width: 275px;
+}
+
+.n-myft-ui__instant-alerts-onboarding__content {
+	margin: 20px 20px 24px;
+}
+
+.n-myft-ui__instant-alerts-onboarding__checkbox {
```
Comment (GlynnPhillips) [inline components/instant-alerts-onboarding/main.scss:16]: **Comment:** I am not sure if we should achieve this in this way. However to be sure I would like to run the demos and play around with it. I am in the middle of doing something in n-myft-ui so I need to find a good stopping point to do that

File: components/instant-alerts-onboarding/instant-alerts-onboarding.html (hunk 3 above)
Comment (asugar13) [reply components/instant-alerts-onboarding/instant-alerts-onboarding.html:2]: will go with `n-myft-ui__preference-modal`

File: components/instant-alerts-onboarding/main.scss (hunk 4 above)
Comment (asugar13) [reply components/instant-alerts-onboarding/main.scss:16]: good call. moving the relative positioning to `n-myft-ui__preferences-modal__content`

File: components/instant-alerts-onboarding/instant-alerts-onboarding.html (hunk 1 above)
Comment (asugar13) [reply components/instant-alerts-onboarding/instant-alerts-onboarding.html]: This is a good suggestion. However, since removing the form passes linting, that's what I've done for now. The main reason is that I think implementing this functionality is part of a separate task to make the API call to change the user's preferences. However, let me know: I could add a simple onClick listener without much functionality to it.

File: components/jsx/instant-alerts-preferences/instant-alerts-preferences.jsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ * Create a popup modal to manage myFT alert preferences
+ * @public
+ * @param {PreferencesProperties}
+ * @returns {React.ReactElement}
+*/
+export default function InstantAlertsPreferencesModal({ flags, currentPreferences }) {
+	//I think in practice this would not be needed 
+	if (!flags.myFtApiWrite) {
+		return null;
+	}
```
Comment (GlynnPhillips) [inline components/jsx/instant-alerts-preferences/instant-alerts-preferences.jsx:21]: **Comment:** Yeah probably not needed but I still think its worth having as its a simply bit of protection just incase someone loads this component without checking first

File: components/jsx/instant-alerts-preferences/instant-alerts-preferences.jsx [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+	//I've created a ticket to handle cases where the user has no preferences set
+	const formattedCurrentPreferences = currentPreferences.join(', ')
```
[the remark this answers is not in the document]
Comment (asugar13) [reply components/jsx/instant-alerts-preferences/instant-alerts-preferences.jsx:24]: sounds good. In this case I added functionality for the fields to be populated dynamically. However there are no designs for the case where there are no alerts set up for the user.

File: components/jsx/instant-alerts-preferences/main.scss [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+	width: 100%;
+	border-radius: 20px;
+	border: 2px solid oColorsByName('black-20');
+	background-color: oColorsByName('white-80');
+}
+
+.n-myft-ui__preferences-modal__remove-button__text {
+	@include oTypographySans($scale: -1, $weight: 'regular', $style: 'normal');
+	color: oColorsByName('black-80');
+}
```
Comment (GlynnPhillips) [inline components/jsx/instant-alerts-preferences/main.scss:46]: **Question:** Do we need this extra child element within the button? If we are only applying font styles then they can be applied directly to the button

File: components/jsx/instant-alerts-preferences/main.scss [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+.n-myft-ui__preferences-modal__manage-preferences {
+	margin-top: 0;
+	margin-bottom: 20px;
+	@include oTypographySans($scale: -1, $weight: 'regular', $style: 'normal');
+}
+
+.n-myft-ui__preferences-modal__remove-button {
+	display: flex;
+	align-items: center;
+	justify-content: center;
```
Comment (GlynnPhillips) [inline components/jsx/instant-alerts-preferences/main.scss:35]: **Comment:** I don't think we need this to be flex, especially if we remove the child element. The horizontal centering can be achieved with `text-align: center` and the vertical centering can be achieved by creating the height with equal padding

File: components/jsx/instant-alerts-preferences/instant-alerts-preferences.jsx [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+							<span class="o-forms-input__label n-myft-ui__preferences-modal__checkbox__message">
+								Get instant alerts for this topic
+							</span>
+						</label>
+					</span>
+
+				<p class="n-myft-ui__preferences-modal__list-preferences">Your alerts are currently: {formattedCurrentPreferences}.</p>
+				<p class="n-myft-ui__preferences-modal__manage-preferences">Manage your preferences <a>here</a></p>
+				<button class="n-myft-ui__preferences-modal__remove-button"><p class="n-myft-ui__preferences-modal__remove-button__text">Remove from myFT</p></button>
```
Comment (GlynnPhillips) [inline components/jsx/instant-alerts-preferences/instant-alerts-preferences.jsx:40]: **Change request:** Buttons can only contain inline elements and `p` is a block level element.

Also see my comment about the css because I think this child element is not required anyway

File: components/jsx/instant-alerts-preferences/main.scss [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+	margin: 20px 20px 24px;
+}
+
+.n-myft-ui__preferences-modal__checkbox__message {
+	@include oTypographySans($scale: 0, $weight: 'semibold', $style: 'normal');
+}
+
+.n-myft-ui__preferences-modal__list-preferences {
+	margin-bottom: 0;
+	font-size: 14px;
```
Comment (GlynnPhillips) [inline components/jsx/instant-alerts-preferences/main.scss:23]: **Comment:** Should we use an origami mixin here to set the font size?

File: components/instant-alerts-preferences/instant-alerts-preferences.html
Comment (GlynnPhillips) [inline components/instant-alerts-preferences/instant-alerts-preferences.html:1]: Do we need a handlebars version?

File: components/jsx/instant-alerts-preferences/instant-alerts-preferences.jsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (GlynnPhillips) [reply components/jsx/instant-alerts-preferences/instant-alerts-preferences.jsx:24]: Yeah I just want to be careful to avoid big bang work. The ticket you are working on calls out that its not inscope to create any of the functionality within the modal. There is already a ticket PT-236 for doing the preference display work

File: components/jsx/instant-alerts-preferences/instant-alerts-preferences.jsx
Comment (GlynnPhillips) [inline components/jsx/instant-alerts-preferences/instant-alerts-preferences.jsx:1]: As we have called the classnames for this to be preference modal should we make the file names match?

File: components/instant-alerts-preferences/instant-alerts-preferences.html
Comment (asugar13) [reply components/instant-alerts-preferences/instant-alerts-preferences.html:1]: whoops! forgot to remove it

File: components/jsx/instant-alerts-preferences/instant-alerts-preferences.jsx
Comment (asugar13) [reply components/jsx/instant-alerts-preferences/instant-alerts-preferences.jsx:1]: I can do that but I felt like one of the reasons to change the classNames was that the original name was a bit wordy. I think the file name gives a bit of a more descriptive explanation of what the component is there for.

Comment (GlynnPhillips) [reply components/jsx/instant-alerts-preferences/instant-alerts-preferences.jsx:1]: > name was a bit wordy

That wasn't a concern for me, I have no issue with wordy names that are descriptive. My concern here is you have a filename that makes it sound very specific but then then code is aimed to make it sound generic. We should pick 1 or the other

Comment (asugar13) [reply components/jsx/instant-alerts-preferences/instant-alerts-preferences.jsx:1]: the original name was `instant-alerts-onboarding` (https://github.com/Financial-Times/n-myft-ui/pull/588#discussion_r1290094114) I will go with `instant-alerts-preferences` instead

Comment (GlynnPhillips) [reply components/jsx/instant-alerts-preferences/instant-alerts-preferences.jsx:1]: > instant-alerts-preferences

I personally think that this is confusing because we have other components that already exist on the site to deliver "instant alerts preferences"

>  I think the file name gives a bit of a more descriptive explanation of what the component is there for


I think the name we choice (preference modal ) is descriptive without being too specific. There is no point in our code being generic so the component can be extended if the filename is then overly specific.

Whatever choice you go with I would like to see the code and the file name reflect each other.

File: components/jsx/instant-alerts-preferences/main.scss (hunk 8 above)
Comment (asugar13) [reply components/jsx/instant-alerts-preferences/main.scss:35]: I believe this was needed when there was a <p> inside the <button> but it's no longer needed:
```    
display: flex;
align-items: center;
justify-content: center;
```

File: components/jsx/preferences-modal/main.scss [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+	height: 32px;
+	width: 100%;
+	border-radius: 20px;
+	border: 2px solid oColorsByName('black-20');
+	background-color: oColorsByName('white-80');
+	@include oTypographySans($scale: -1, $weight: 'regular', $style: 'normal');
+	color: oColorsByName('black-80');
+}
+
+.n-myft-ui__preferences-modal__remove-button:hover {
```
Comment (GlynnPhillips) [inline components/jsx/preferences-modal/main.scss:42]: **Change request:** Here I think we need to account for `focus` state as well to allow for keyboard navigation

```suggestion
.n-myft-ui__preferences-modal__remove-button:hover,
.n-myft-ui__preferences-modal__remove-button:focus {
```

File: components/jsx/preferences-modal/preferences-modal.jsx [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+								Get instant alerts for this topic
+							</span>
+						</label>
+					</span>
+
+				<p class="n-myft-ui__preferences-modal__list-preferences">Your alerts are currently: {formattedCurrentPreferences}.</p>
+				<p class="n-myft-ui__preferences-modal__manage-preferences">Manage your preferences <a href="/myft/alerts">here</a></p>
+				<button class="n-myft-ui__preferences-modal__remove-button">Remove from myFT</button>
+			</div>
+		</div>
```
Comment (GlynnPhillips) [inline components/jsx/preferences-modal/preferences-modal.jsx:40]: **Change request:** This is now jsx so all the `class` attributes in here need to be `className`

File: components/jsx/preferences-modal/preferences-modal.jsx [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+		return null;
+	}
+
+	const formattedCurrentPreferences = currentPreferences.join(', ')
+	return (
+		<div class="n-myft-ui__preferences-modal">
+			<div class="n-myft-ui__preferences-modal__content">
+					<span class="o-forms-input o-forms-input--checkbox">
+						<label for="receive-instant-alerts">
+							<input type="checkbox" name="receive-instant-alerts" value="receive-instant-alerts"></input>
```
Comment (GlynnPhillips) [inline components/jsx/preferences-modal/preferences-modal.jsx:28]: **Nitpick:** `input` can be self closing here if you want 

```suggestion
							<input type="checkbox" name="receive-instant-alerts" value="receive-instant-alerts" />
```

File: components/jsx/preferences-modal/preferences-modal.jsx [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+							<span class="o-forms-input__label n-myft-ui__preferences-modal__checkbox__message">
+								Get instant alerts for this topic
+							</span>
+						</label>
+					</span>
+
+				<p class="n-myft-ui__preferences-modal__list-preferences">Your alerts are currently: {formattedCurrentPreferences}.</p>
+				<p class="n-myft-ui__preferences-modal__manage-preferences">Manage your preferences <a href="/myft/alerts">here</a></p>
```
Comment (GlynnPhillips) [inline components/jsx/preferences-modal/preferences-modal.jsx:37]: **Comment/Change request:** Maybe this can't be a change request until we speak to Design or Product but we should avoid "here" or "click here" type of links as they are bad for accessibility. For example if a user is use assistive tech to navigate via links in the page they wont know what "here" is as they wont have the "manage your preference" context.

I would suggest we do something like the below 
```suggestion
				<a href="/myft/alerts" className="n-myft-ui__preferences-modal__manage-preferences">Manage your preferences </a>
```

**Review by GlynnPhillips — CHANGES_REQUESTED** [review]
A couple of minor improvements but its looking good

File: components/jsx/preferences-modal/preferences-modal.jsx (hunk 13 above)
Comment (asugar13) [reply components/jsx/preferences-modal/preferences-modal.jsx:28]: true. it was self-closing at some point in the review. Can't remember why I added the closing tag.

File: components/jsx/preferences-modal/preferences-modal.jsx (hunk 14 above)
Comment (asugar13) [reply components/jsx/preferences-modal/preferences-modal.jsx:37]: Instead of committing your suggestion, should I add it to this ticket to discuss with designer once he's back from time off? : https://financialtimes.atlassian.net/browse/PT-249

Comment (asugar13) [reply components/jsx/preferences-modal/preferences-modal.jsx]: I've gone ahead and made that change and updated the ticket so we can run it by Mark when he's back.

**Review by GlynnPhillips — APPROVED** [review]

**Comment by gyss** [comment]
Looks good @asugar13 ! just one thing, on mobile, looks like the component is taking more width than the grid container. Is this being considered?

Screenshot 2023-08-16 at 11 39 01

File: components/jsx/preferences-modal/preferences-modal.jsx [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+		return null;
+	}
+
+	const formattedCurrentPreferences = currentPreferences.join(', ')
+	return (
+		<div className="n-myft-ui__preferences-modal">
+			<div className="n-myft-ui__preferences-modal__content">
+					<span className="o-forms-input o-forms-input--checkbox">
+						<label htmlFor="receive-instant-alerts">
+							<input type="checkbox" name="receive-instant-alerts" value="receive-instant-alerts"/>
```
Comment (metart43) [inline components/jsx/preferences-modal/preferences-modal.jsx:28]: I might be wrong but I reckon the association between `<label>` and `<input>` is done through the `id` prop.

So, you just simply need to add `id` property to the `<input>` checkbox

```jsx
<input id="receive-instant-alerts" type="checkbox" name="receive-instant-alerts" value="receive-instant-alerts" />
```

File: components/jsx/preferences-modal/preferences-modal.jsx [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+							<span className="o-forms-input__label n-myft-ui__preferences-modal__checkbox__message">
+								Get instant alerts for this topic
+							</span>
+						</label>
+					</span>
+
+				<p className="n-myft-ui__preferences-modal__text">Your alerts are currently: {formattedCurrentPreferences}.</p>
+				<a className="n-myft-ui__preferences-modal__text">Manage your preferences here</a>
```
Comment (metart43) [inline components/jsx/preferences-modal/preferences-modal.jsx:36]: Sorry, I am not sure why there is an `<a>` tag with a `href` property.


Without a valid `href`, this element could be confusing to screen readers and other assistive technologies. If it's meant to be interactive (like a button), it's better to use a `<button>` element or provide appropriate roles and ARIA attributes.

**Review by metart43 — CHANGES_REQUESTED** [review]

File: components/jsx/preferences-modal/preferences-modal.jsx (hunk 15 above)
Comment (asugar13) [reply components/jsx/preferences-modal/preferences-modal.jsx:28]: Sounds good, I'll add the `id`. It seems like it's a good accessibility practice

File: components/jsx/preferences-modal/preferences-modal.jsx (hunk 16 above)
Comment (asugar13) [reply components/jsx/preferences-modal/preferences-modal.jsx:36]: Good spot! I think the href went away in a series of rebases I did.

File: components/jsx/preferences-modal/preferences-modal.jsx (hunk 15 above)
Comment (GlynnPhillips) [reply components/jsx/preferences-modal/preferences-modal.jsx]: Just for future reference, if you have an implicit label & input relationship (eg. the label is wrapping the input) then you don't need the `for` & `id` relationship (explicit relationship) - however doing both is also not a problem

Comment (GlynnPhillips) [reply components/jsx/preferences-modal/preferences-modal.jsx]: Explicit relationships are better for accessibility but they still don't work perfectly if combined with an implicit relationship. However our styling via forms does rely on an implicit relationship so I don't think theres a way around that

**Review by asugar13 — COMMENTED** [review]
@gyss , this is how the component looks like on mobile when being imported into other apps such as `next-article`: 
Screenshot 2023-08-16 at 13 53 12

**Review by gyss — APPROVED** [review]

**Review by metart43 — APPROVED** [review]
